# Fase 1: Seleccion y reconocimiento de datos

In [4]:
#libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pathlib as pl
from libmac import listar_archivos, etiquetar_unirdatosdeclase

## 1. Dataset del autor (flujos presentados)

In [3]:
#import data
dfa_normal = pd.read_csv('InSDN_DatasetCSV/Normal_data.csv')
dfa_ataques = pd.read_csv('InSDN_DatasetCSV/OVS.csv')

### 1.1 Tráfico normal

In [4]:
#show features db normal
dfa_normal.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68424 entries, 0 to 68423
Data columns (total 84 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Flow ID            68424 non-null  object 
 1   Src IP             68424 non-null  object 
 2   Src Port           68424 non-null  int64  
 3   Dst IP             68424 non-null  object 
 4   Dst Port           68424 non-null  int64  
 5   Protocol           68424 non-null  int64  
 6   Timestamp          68424 non-null  object 
 7   Flow Duration      68424 non-null  int64  
 8   Tot Fwd Pkts       68424 non-null  int64  
 9   Tot Bwd Pkts       68424 non-null  int64  
 10  TotLen Fwd Pkts    68424 non-null  float64
 11  TotLen Bwd Pkts    68424 non-null  float64
 12  Fwd Pkt Len Max    68424 non-null  int64  
 13  Fwd Pkt Len Min    68424 non-null  int64  
 14  Fwd Pkt Len Mean   68424 non-null  float64
 15  Fwd Pkt Len Std    68424 non-null  float64
 16  Bwd Pkt Len Max    684

In [5]:
#show statics of db normal
dfa_normal.describe()

Src Port      Dst Port      Protocol  Flow Duration  Tot Fwd Pkts  \
count  68424.000000  68424.000000  68424.000000   6.842400e+04  68424.000000   
mean   37361.032459   9733.738557     11.418947   1.362636e+07     11.807217   
std    20305.851419  19352.308798      5.551915   3.360107e+07    112.712183   
min        0.000000      0.000000      0.000000  -1.540000e+02      0.000000   
25%    34213.500000     53.000000      6.000000   3.134000e+03      1.000000   
50%    43383.000000    443.000000      6.000000   4.799000e+03      1.000000   
75%    52105.250000    443.000000     17.000000   7.122470e+05      6.000000   
max    65513.000000  60998.000000     17.000000   1.200000e+08  16928.000000   

       Tot Bwd Pkts  TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  \
count  68424.000000     6.842400e+04     6.842400e+04     68424.000000   
mean      18.592628     3.354497e+03     4.128116e+04       425.863644   
std      235.949827     1.560998e+05     7.680108e+05      1425.400448   
min        1.000000     0.000000e+00     0.000000e+00         0.000000   
25%        2.000000     3.000000e+01     1.370000e+02        30.000000   
50%        3.000000     3.600000e+01     5.020000e+02        36.000000   
75%        5.000000     5.990000e+02     7.620000e+02       249.000000   
max    34094.000000     3.160000e+07     1.070000e+08     64239.000000   

       Fwd Pkt Len Min  ...  Fwd Act Data Pkts  Fwd Seg Size Min  \
count     68424.000000  ...       68424.000000           68424.0   
mean         22.297469  ...           3.891266               0.0   
std          67.228213  ...          57.082340               0.0   
min           0.000000  ...           0.000000               0.0   
25%           0.000000  ...           1.000000               0.0   
50%           0.000000  ...           1.000000               0.0   
75%          34.000000  ...           3.000000               0.0   
max        3900.000000  ...       11180.000000               0.0   

        Active Mean    Active Std    Active Max    Active Min     Idle Mean  \
count  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04   
mean   1.887697e+05  1.434025e+05  3.980718e+05  8.563484e+04  4.508770e+06   
std    1.373001e+06  1.105629e+06  2.467808e+06  1.027435e+06  1.314630e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.040000e+08  6.830000e+07  1.040000e+08  1.040000e+08  1.190000e+08   

           Idle Std      Idle Max      Idle Min  
count  6.842400e+04  6.842400e+04  6.842400e+04  
mean   6.041184e+05  5.066966e+06  3.975015e+06  
std    3.486207e+06  1.448106e+07  1.246315e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.000000e+00  
75%    0.000000e+00  0.000000e+00  0.000000e+00  
max    7.300000e+07  1.190000e+08  1.190000e+08  

[8 rows x 79 columns]

In [6]:
#encontrar tipos de protocolos
dfa_normal.Protocol.unique()

array([ 6,  0, 17], dtype=int64)

In [7]:
#tipo de trafico
dfa_normal.Label.unique()

array(['Normal'], dtype=object)

**Descripción subconjunto de tráfico normal:**<br>
El subconjunto de tráfico normal cuenta con un total de 84 caracteríticas y 68424 muestras. Los registros de flujo analizados contiene tres tipos de datos: object, float e int. Las caracteristicas de tipo object son: 
- Flow ID (contiene la IP de origen y destino, puerto de origen y destino, y protocolo L4).
- Src IP (direccion ip de origen).
- Dst IP (direccion ip de destino). 
- Timestamp: marca de tiempo.
- Label: etiqueta.

La caracteristica Protocolo es de tipo Int y contiene 3 valores; HOPOPT, UDP, TCP, además estos datos ya se encuentran enumerados.
También, existen conjuntos de atributos estadísticos referentes a los flujos de tráficos, intervalos de tiempo, bytes, paquetes, entre otros. Estos últimos se analizarán con más detalle en la siguiente fase.

### 1.2 Tráfico de ataque

In [10]:
#mostrar caracteristicas de los datos de ataques
dfa_ataques.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138722 entries, 0 to 138721
Data columns (total 84 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Flow ID            138722 non-null  object 
 1   Src IP             138722 non-null  object 
 2   Src Port           138722 non-null  int64  
 3   Dst IP             138722 non-null  object 
 4   Dst Port           138722 non-null  int64  
 5   Protocol           138722 non-null  int64  
 6   Timestamp          138722 non-null  object 
 7   Flow Duration      138722 non-null  int64  
 8   Tot Fwd Pkts       138722 non-null  int64  
 9   Tot Bwd Pkts       138722 non-null  int64  
 10  TotLen Fwd Pkts    138722 non-null  int64  
 11  TotLen Bwd Pkts    138722 non-null  int64  
 12  Fwd Pkt Len Max    138722 non-null  int64  
 13  Fwd Pkt Len Min    138722 non-null  int64  
 14  Fwd Pkt Len Mean   138722 non-null  float64
 15  Fwd Pkt Len Std    138722 non-null  float64
 16  Bw

In [9]:
#mostrar estadisticas del trafico de ataques
dfa_ataques.describe()

Src Port      Dst Port      Protocol  Flow Duration  Tot Fwd Pkts  \
count  68424.000000  68424.000000  68424.000000   6.842400e+04  68424.000000   
mean   37361.032459   9733.738557     11.418947   1.362636e+07     11.807217   
std    20305.851419  19352.308798      5.551915   3.360107e+07    112.712183   
min        0.000000      0.000000      0.000000  -1.540000e+02      0.000000   
25%    34213.500000     53.000000      6.000000   3.134000e+03      1.000000   
50%    43383.000000    443.000000      6.000000   4.799000e+03      1.000000   
75%    52105.250000    443.000000     17.000000   7.122470e+05      6.000000   
max    65513.000000  60998.000000     17.000000   1.200000e+08  16928.000000   

       Tot Bwd Pkts  TotLen Fwd Pkts  TotLen Bwd Pkts  Fwd Pkt Len Max  \
count  68424.000000     6.842400e+04     6.842400e+04     68424.000000   
mean      18.592628     3.354497e+03     4.128116e+04       425.863644   
std      235.949827     1.560998e+05     7.680108e+05      1425.400448   
min        1.000000     0.000000e+00     0.000000e+00         0.000000   
25%        2.000000     3.000000e+01     1.370000e+02        30.000000   
50%        3.000000     3.600000e+01     5.020000e+02        36.000000   
75%        5.000000     5.990000e+02     7.620000e+02       249.000000   
max    34094.000000     3.160000e+07     1.070000e+08     64239.000000   

       Fwd Pkt Len Min  ...  Fwd Act Data Pkts  Fwd Seg Size Min  \
count     68424.000000  ...       68424.000000           68424.0   
mean         22.297469  ...           3.891266               0.0   
std          67.228213  ...          57.082340               0.0   
min           0.000000  ...           0.000000               0.0   
25%           0.000000  ...           1.000000               0.0   
50%           0.000000  ...           1.000000               0.0   
75%          34.000000  ...           3.000000               0.0   
max        3900.000000  ...       11180.000000               0.0   

        Active Mean    Active Std    Active Max    Active Min     Idle Mean  \
count  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04  6.842400e+04   
mean   1.887697e+05  1.434025e+05  3.980718e+05  8.563484e+04  4.508770e+06   
std    1.373001e+06  1.105629e+06  2.467808e+06  1.027435e+06  1.314630e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.040000e+08  6.830000e+07  1.040000e+08  1.040000e+08  1.190000e+08   

           Idle Std      Idle Max      Idle Min  
count  6.842400e+04  6.842400e+04  6.842400e+04  
mean   6.041184e+05  5.066966e+06  3.975015e+06  
std    3.486207e+06  1.448106e+07  1.246315e+07  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  0.000000e+00  
75%    0.000000e+00  0.000000e+00  0.000000e+00  
max    7.300000e+07  1.190000e+08  1.190000e+08  

[8 rows x 79 columns]

In [11]:
#encontrar tipos de protocolos
dfa_ataques.Protocol.unique()

array([ 6,  0, 17], dtype=int64)

In [12]:
# tipos de ataques
dfa_ataques.Label.unique()

array(['BFA', 'DDoS ', 'DoS', 'Probe', 'Web-Attack', 'BOTNET'],
      dtype=object)

**Descripción subconjunto de tráfico de ataque:**<br>
El subconjunto de tráfico de ataque cuenta con 84 caracteríticas y 138722 muestras. Los tipos de ataque son: fuerza bruta (BFA), DDoS, Dos, Probe, Web y botnet, cuyas muestras están desequilibradas. Así mismo, los registros de flujo contiene tres tipos de datos: object, float e int. Las caracteristicas de tipo object siguen siendo las mismas del anterior (normal). Se repiten los mismos protocolos codificados de L4 HOPOPT, UDP y TCP.
También, existen los mismos conjuntos de atributos estadísticos referentes a los flujos de tráficos, intervalos de tiempo, bytes, paquetes, etc, los cuales serán analizados con más detalle en la siguiente fase.

### 1.3 Total de datos

In [13]:
#concatenar dataset
df_unido = pd.concat([dfa_normal,dfa_ataques], axis=0)
#dimension del conjunto total 
df_unido.shape

(207146, 84)

In [15]:
#seleccion de variable independiente
X = df_unido.iloc[:,:-1].values
#seleccion de variable dependiente u objetivo
y=df_unido.iloc[:,83]

In [16]:
#cantidad de muestras por clase
print(f"{y.value_counts()}",f"{y.value_counts().sum()}" )

Normal        68424
DoS           52471
DDoS          48413
Probe         36372
BFA            1110
Web-Attack      192
BOTNET          164
Name: Label, dtype: int64 207146


# 2. Datos extraídos con cicflowmeter version Java

In [2]:
#diccionario de acceso a archivos de extraccion de flujos
dir_trafico = pl.Path("/ML-Tesis-Attacks/PcapToCsv/windows/trafico/")
"""crear diccionario para el acceso de archivos"""
dicc_traficoJ = {
    'Normal':list(dir_trafico.glob('normal/*.csv')),
    'BOTNET': list(dir_trafico.glob('Botnet/*.csv')),
    'BFA': list(dir_trafico.glob('brute-force-attack/*.csv')),  
    'DDoS': list(dir_trafico.glob('DDoS/*.csv')),  
    'Dos': list(dir_trafico.glob('dos/**/*.csv')),  
    'Probe': list(dir_trafico.glob('Probe/*.csv')),  
    'Web-Attack': list(dir_trafico.glob('web_attack/*.csv'))  
}

In [3]:
#mostrar nro de archivos
listar_archivos(dicc_traficoJ)

Clase: Normal
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
Clase: BOTNET
1
Clase: BFA
1
2
3
4
Clase: DDoS
1
2
3
Clase: Dos
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
Clase: Probe
1
2
3
4
Clase: Web-Attack
1
2
3
4
5
6


In [4]:
#unir datos extraidos de la misma clase y etiquetarlos
etiquetar_unirdatosdeclase(dicc_traficoJ, 'Label', '/ML-Tesis-Attacks/datos_SDN/claseDatosJW/')

En proceso de clase Normal...
Proceso completado.... 
 Total de archivos unidos: 22
En proceso de clase BOTNET...
Proceso completado.... 
 Total de archivos unidos: 1
En proceso de clase BFA...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase DDoS...
Proceso completado.... 
 Total de archivos unidos: 3
En proceso de clase Dos...
Proceso completado.... 
 Total de archivos unidos: 18
En proceso de clase Probe...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase Web-Attack...
Proceso completado.... 
 Total de archivos unidos: 6


In [5]:
#importar
df_bfajw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/BFA.csv')
df_botnetjw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/BOTNET.csv')
df_ddosjw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/DDoS.csv')
df_dosjw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/Dos.csv')
df_normaljw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/Normal.csv')
df_probejw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/Probe.csv')
df_webattackjw=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosJW/Web-Attack.csv')

In [6]:
#concatenar dataset
df_jw = pd.concat([df_bfajw, df_botnetjw, df_ddosjw, df_dosjw, df_normaljw, df_probejw, df_webattackjw], axis=0)

In [7]:
#cantidad de muestras por clase
df_jw.Label.value_counts()

Dos           69651
Normal        60114
Probe         36723
BFA            1136
Web-Attack      204
BOTNET          164
DDoS             25
Name: Label, dtype: int64

# 3. Datos extraídos con cicflowmeter version Python

In [3]:
#diccionario de acceso a archivos de extraccion de flujos
dir_traficoP = pl.Path("/ML-Tesis-Attacks/PcapToCsv/ubuntu/trafico/")
"""crear diccionario para el acceso de archivos"""
dicc_traficoP = {
    'Normal':list(dir_traficoP.glob('normal/**/*.csv')),
    'BOTNET': list(dir_traficoP.glob('Botnet/*.csv')),
    'BFA': list(dir_traficoP.glob('brute-force-attack/*.csv')),  
    'DDoS': list(dir_traficoP.glob('DDoS/*.csv')),  
    'Dos': list(dir_traficoP.glob('dos/**/*.csv')),  
    'Probe': list(dir_traficoP.glob('Probe/*.csv')),  
    'Web-Attack': list(dir_traficoP.glob('web_attack/*.csv'))  
}

In [4]:
listar_archivos(dicc_traficoP)

Clase: Normal
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
Clase: BOTNET
1
2
Clase: BFA
1
2
3
4
Clase: DDoS
1
2
3
4
Clase: Dos
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
Clase: Probe
1
2
3
4
Clase: Web-Attack
1
2
3
4
5
6


In [5]:
#unir datos por clase
etiquetar_unirdatosdeclase(dicc_traficoP,'Label', '/ML-Tesis-Attacks/datos_SDN/claseDatosPU/')

En proceso de clase Normal...
Proceso completado.... 
 Total de archivos unidos: 22
En proceso de clase BOTNET...
Proceso completado.... 
 Total de archivos unidos: 2
En proceso de clase BFA...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase DDoS...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase Dos...
Proceso completado.... 
 Total de archivos unidos: 18
En proceso de clase Probe...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase Web-Attack...
Proceso completado.... 
 Total de archivos unidos: 6


In [6]:
#importar archivos de datos generados
df_bfapu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/BFA.csv')
df_botnetpu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/BOTNET.csv')
df_ddospu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/DDoS.csv')
df_dospu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/Dos.csv')
df_normalpu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/Normal.csv')
df_probepu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/Probe.csv')
df_webattackpu=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/claseDatosPU/Web-Attack.csv')

In [7]:
#concatenar dataset
df_pu = pd.concat([df_bfapu, df_botnetpu, df_ddospu, df_dospu, df_normalpu, df_probepu, df_webattackpu], axis=0)

In [8]:
#cantidad de muestras por clase
df_pu.Label.value_counts()

DDoS          398643
Normal         53784
Probe          41980
Dos            33842
BFA              582
Web-Attack       120
BOTNET            92
Name: Label, dtype: int64

# 4. Extraccion de datos con Argus

In [2]:
#diccionario de acceso a archivos de extraccion de flujos
dir_traficoAr = pl.Path("/ML-Tesis-Attacks/argus_espec/")
"""crear diccionario para el acceso de archivos"""
dicc_traficoAr = {
    'Normal':list(dir_traficoAr.glob('normal/**/*.csv')),  
    'Botnet': list(dir_traficoAr.glob('Botnet/*.csv')),  
    'Dos': list(dir_traficoAr.glob('dos/**/*.csv')),  
    'Probe': list(dir_traficoAr.glob('Probe/*.csv'))  
}

In [3]:
etiquetar_unirdatosdeclase(dicc_traficoAr, 'Label', '/ML-Tesis-Attacks/datos_SDN/argDatos/')

En proceso de clase Normal...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN\Fase1\libmac.py:37: DtypeWarning: Columns (113) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN\Fase1\libmac.py:37: DtypeWarning: Columns (113) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN\Fase1\libmac.py:37: DtypeWarning: Columns (113) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN\Fase1\libmac.py:37: DtypeWarning: Columns (113) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)


Proceso completado.... 
 Total de archivos unidos: 22
En proceso de clase Botnet...
Proceso completado.... 
 Total de archivos unidos: 2
En proceso de clase Dos...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN\Fase1\libmac.py:37: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN\Fase1\libmac.py:37: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)


Proceso completado.... 
 Total de archivos unidos: 18
En proceso de clase Probe...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN\Fase1\libmac.py:37: DtypeWarning: Columns (27,35,112,113) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)


Proceso completado.... 
 Total de archivos unidos: 4


In [3]:
#importar archivos de datos generados
df_botA=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/argDatos/Botnet.csv')
df_dosA=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/argDatos/Dos.csv')
df_normalA=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/argDatos/Normal.csv')
df_probeA=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/argDatos/Probe.csv')

C:\Users\miancastillo\AppData\Local\Temp\ipykernel_21728\1916635651.py:3: DtypeWarning: Columns (22,113) have mixed types. Specify dtype option on import or set low_memory=False.
  df_dosA=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/argDatos/Dos.csv')
C:\Users\miancastillo\AppData\Local\Temp\ipykernel_21728\1916635651.py:4: DtypeWarning: Columns (113) have mixed types. Specify dtype option on import or set low_memory=False.
  df_normalA=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/argDatos/Normal.csv')
C:\Users\miancastillo\AppData\Local\Temp\ipykernel_21728\1916635651.py:5: DtypeWarning: Columns (22,113) have mixed types. Specify dtype option on import or set low_memory=False.
  df_probeA=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/argDatos/Probe.csv')


In [5]:
#concatenar datasets
df_a = pd.concat([ df_botA, df_dosA, df_normalA, df_probeA], axis=0)

In [6]:
df_a.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 188674 entries, 0 to 51665
Data columns (total 123 columns):
 #    Column       Dtype  
---   ------       -----  
 0    SrcId        object 
 1    Rank         float64
 2    StartTime    object 
 3    LastTime     object 
 4    Trans        float64
 5    Flgs         object 
 6    Seq          int64  
 7    Dur          float64
 8    RunTime      float64
 9    IdleTime     float64
 10   Mean         float64
 11   StdDev       float64
 12   Sum          float64
 13   Min          float64
 14   Max          float64
 15   SrcMac       object 
 16   DstMac       object 
 17   SrcOui       float64
 18   DstOui       float64
 19   SrcAddr      object 
 20   DstAddr      object 
 21   Proto        object 
 22   Sport        object 
 23   Dport        object 
 24   sTos         float64
 25   dTos         float64
 26   sDSb         object 
 27   dDSb         object 
 28   sCo          float64
 29   dCo          float64
 30   sTtl         float6

In [22]:
#importa por separado archivos DDoS
df_d1=pd.read_csv('/ML-Tesis-Attacks/argus_2/DDoS/fa-ONOS-DDoS-ICMP.csv')
df_d2=pd.read_csv('/ML-Tesis-Attacks/argus_2/DDoS/fa-ONOS-DDoS-SYN.csv')
df_d3=pd.read_csv('/ML-Tesis-Attacks/argus_2/DDoS/fa-ONOS-DDoS-UDP.csv')
df_d4=pd.read_csv('/ML-Tesis-Attacks/argus_2/DDoS/fa-h4-DDoS-ICMP.csv')
df_d5=pd.read_csv('/ML-Tesis-Attacks/argus_2/DDoS/fa-h4-DDoS-SYN.csv')
df_d6=pd.read_csv('/ML-Tesis-Attacks/argus_2/DDoS/fa-h4-DDoS-UDP.csv')

C:\Users\miancastillo\AppData\Local\Temp\ipykernel_22116\977376132.py:4: DtypeWarning: Columns (27,35) have mixed types. Specify dtype option on import or set low_memory=False.
  df_d4=pd.read_csv('/ML-Tesis-Attacks/argus_2/DDoS/fa-h4-DDoS-ICMP.csv')
C:\Users\miancastillo\AppData\Local\Temp\ipykernel_22116\977376132.py:5: DtypeWarning: Columns (22) have mixed types. Specify dtype option on import or set low_memory=False.
  df_d5=pd.read_csv('/ML-Tesis-Attacks/argus_2/DDoS/fa-h4-DDoS-SYN.csv')


In [3]:
df_d1

SrcId  Rank        StartTime         LastTime  Trans        Flgs  Seq  \
0   0.0.0.0   NaN  12:24:23.274088  12:24:23.266110    NaN                0   
1   0.0.0.0   1.0  10:20:45.697407  10:20:45.697407    1.0   e            1   
2   0.0.0.0   2.0  10:20:45.697867  10:22:03.366429    1.0   e *@         2   
3   0.0.0.0   3.0  10:20:45.697882  10:21:53.693519    1.0   e            3   
4   0.0.0.0   4.0  10:20:45.697886  10:21:53.824121    1.0   e *          4   
5   0.0.0.0   5.0  10:21:37.317502  10:21:37.617097    1.0   e     A      5   
6   0.0.0.0   6.0  10:21:37.317514  10:21:37.317514    1.0   e     A      6   
7   0.0.0.0   7.0  10:21:37.341897  10:21:37.341897    1.0   e     A      7   
8   0.0.0.0   8.0  10:21:37.342495  10:21:37.342495    1.0   e     A      8   
9   0.0.0.0   9.0  10:21:37.342981  10:21:37.342981    1.0   e     A      9   
10  0.0.0.0  10.0  10:21:37.344577  10:21:37.359571    1.0   e           10   
11  0.0.0.0  11.0  10:21:37.347127  10:21:37.356283    1.0   e           11   
12  0.0.0.0  12.0  10:21:37.351545  10:21:52.364580    1.0   e           12   
13  0.0.0.0  13.0  10:21:37.356270  10:21:37.356270    1.0   e           13   
14  0.0.0.0  14.0  10:21:37.356293  10:21:37.356293    1.0   e           14   
15  0.0.0.0  15.0  10:21:37.616972  10:21:37.616972    1.0   e     A     15   
16  0.0.0.0  16.0  10:21:52.456953  10:21:54.366382    1.0   e           16   
17  0.0.0.0  17.0  10:21:54.453754  10:21:55.494735    1.0   e           17   
18  0.0.0.0  18.0  10:21:55.453619  10:21:55.736960    1.0   e           18   
19  0.0.0.0  19.0  10:21:57.453660  10:21:59.495024    1.0   e           19   
20  0.0.0.0  20.0  10:21:59.453585  10:21:59.500730    1.0   e           20   
21  0.0.0.0  20.0  12:24:24.284465  12:24:23.266110    NaN                0   

          Dur    RunTime      IdleTime  ...  Inode  Offset   sMeanPktSz  \
0    0.007978        NaN           NaN  ...    NaN       0     0.000000   
1    0.000000   0.000000  1.672696e+09  ...    NaN     128   216.000000   
2   77.668564  77.668564  1.672696e+09  ...    NaN     240  1165.389038   
3   67.995636  67.995636  1.672696e+09  ...    NaN     600   837.449524   
4   68.126236  68.126236  1.672696e+09  ...    NaN     960   709.027771   
5    0.299595   0.299595  1.672696e+09  ...    NaN    1320    90.000000   
6    0.000000   0.000000  1.672696e+09  ...    NaN    1516    60.000000   
7    0.000000   0.000000  1.672696e+09  ...    NaN    1632    60.000000   
8    0.000000   0.000000  1.672696e+09  ...    NaN    1748    60.000000   
9    0.000000   0.000000  1.672696e+09  ...    NaN    1864    60.000000   
10   0.014994   0.014994  1.672696e+09  ...    NaN    1980    99.000000   
11   0.009156   0.009156  1.672696e+09  ...    NaN    2124   119.000000   
12  15.013035  15.013035  1.672696e+09  ...    NaN    2292   179.000000   
13   0.000000   0.000000  1.672696e+09  ...    NaN    2436    94.000000   
14   0.000000   0.000000  1.672696e+09  ...    NaN    2572    74.000000   
15   0.000000   0.000000  1.672696e+09  ...    NaN    2684    60.000000   
16   1.909429   1.909429  1.672696e+09  ...    NaN    2800    65.714287   
17   1.040981   1.040981  1.672696e+09  ...    NaN    3128    70.000000   
18   0.283341   0.283341  1.672696e+09  ...    NaN    3436   175.666672   
19   2.041364   2.041364  1.672696e+09  ...    NaN    3784    70.000000   
20   0.047145   0.047145  1.672696e+09  ...    NaN    4092   175.666672   
21   1.018355        NaN           NaN  ...    NaN    4092     0.000000   

    dMeanPktSz  sPktSz sMaxPktSz dPktSz  dMaxPktSz  sMinPktSz dMinPktSz  
0     0.000000     NaN       NaN    NaN        NaN        NaN       NaN  
1     0.000000     NaN     216.0    NaN        NaN      216.0       NaN  
2   262.508453     NaN   65226.0    NaN    65226.0       66.0      66.0  
3   127.814369     NaN    6178.0    NaN     1143.0       60.0      66.0  
4   127.238098     NaN    6178.0    NaN      322.0       60.0      66.0  
5     0.000000    

In [4]:
df_d1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Columns: 123 entries, SrcId to dMinPktSz
dtypes: float64(86), int64(18), object(19)
memory usage: 21.3+ KB


In [5]:
df_d1.TotPkts.unique()

array([     0,      1, 116601,    276,    255,      5,      4,      6,
           13,      7,     25,     23], dtype=int64)

In [6]:
df_d1.TotBytes.unique()

array([       0,      216, 49008553,   112627,    95279,      450,
             60,      396,      476,     1074,       94,       74,
            888,      486,     3074,     2942, 10965204], dtype=int64)

In [7]:
df_d1.Dur.unique()

array([7.9780000e-03, 0.0000000e+00, 7.7668564e+01, 6.7995636e+01,
       6.8126236e+01, 2.9959500e-01, 1.4994000e-02, 9.1560000e-03,
       1.5013035e+01, 1.9094290e+00, 1.0409810e+00, 2.8334100e-01,
       2.0413640e+00, 4.7145000e-02, 1.0183550e+00])

In [9]:
df_d1.Mean.unique()

array([          nan, 0.0000000e+00, 7.7668564e+01, 6.7995636e+01,
       6.8126236e+01, 2.9959500e-01, 1.4994000e-02, 9.1560000e-03,
       1.5013035e+01, 1.9094290e+00, 1.0409810e+00, 2.8334100e-01,
       2.0413640e+00, 4.7145000e-02])

In [10]:
df_d1.StdDev.unique()

array([nan,  0.])

In [11]:
df_d1.Sum.unique()

array([          nan, 0.0000000e+00, 7.7668564e+01, 6.7995636e+01,
       6.8126236e+01, 2.9959500e-01, 1.4994000e-02, 9.1560000e-03,
       1.5013035e+01, 1.9094290e+00, 1.0409810e+00, 2.8334100e-01,
       2.0413640e+00, 4.7145000e-02])

In [12]:
df_d1.Min.unique()

array([          nan, 0.0000000e+00, 7.7668564e+01, 6.7995636e+01,
       6.8126236e+01, 2.9959500e-01, 1.4994000e-02, 9.1560000e-03,
       1.5013035e+01, 1.9094290e+00, 1.0409810e+00, 2.8334100e-01,
       2.0413640e+00, 4.7145000e-02])

In [13]:
df_d1.Max.unique()

array([          nan, 0.0000000e+00, 7.7668564e+01, 6.7995636e+01,
       6.8126236e+01, 2.9959500e-01, 1.4994000e-02, 9.1560000e-03,
       1.5013035e+01, 1.9094290e+00, 1.0409810e+00, 2.8334100e-01,
       2.0413640e+00, 4.7145000e-02])

In [14]:
df_d1.SrcPkts.unique()

array([    0,     1, 20379,   109,   108,     5,     4,     6,     7,
          12], dtype=int64)

In [15]:
df_d1.DstPkts.unique()

array([    0, 96222,   167,   147,     6,     3,    13,    11,    21],
      dtype=int64)

In [16]:
df_d1.SrcBytes.unique()

array([       0,      216, 23749465,    91282,    76575,      450,
             60,      396,      476,     1074,       94,       74,
            460,      280,     2108], dtype=int64)

In [17]:
df_d1.DstBytes.unique()

array([       0, 25259088,    21345,    18704,      428,      206,
            966,      834], dtype=int64)

In [19]:
df_d1.Rate.unique()

array([0.00000000e+00, 1.50125085e+03, 4.04437700e+00, 3.72837300e+00,
       1.33513570e+01, 2.00080032e+02, 3.27653992e+02, 3.33044000e-01,
       6.28460200e+00, 5.76379300e+00, 8.47035900e+01, 2.93921100e+00,
       4.66645447e+02])

In [20]:
df_d1.SrcRate.unique()

array([  0.      , 262.424591,   1.588337,   1.570614,  13.351357,
       200.080032, 327.653992,   0.333044,   3.142301,   3.000957,
        38.825356,   1.500106, 233.476257])

In [21]:
df_d1.DstRate.unique()

array([0.00000000e+00, 1.23888611e+03, 2.44138100e+00, 2.14312800e+00,
       2.61962000e+00, 1.92255200e+00, 4.23535880e+01, 9.79748000e-01,
       2.12197098e+02])

In [23]:
df_ddost=pd.concat([df_d1,df_d2,df_d3,df_d4,df_d5,df_d6],axis=0)

In [26]:
df_ddost.Label= 'DDoS'

In [6]:
#mostrar total de muestras por clase
df_a.Label.value_counts()

Normal    101066
Probe      51666
Dos        35810
Botnet       132
Name: Label, dtype: int64

In [27]:
df_ddost.Label.value_counts()

DDoS    5847197
Name: Label, dtype: int64

In [13]:
df_b1=pd.read_csv('/ML-Tesis-Attacks/argus_bots/fa-2017-06-24_win3emotet.csv')
df_b2=pd.read_csv('/ML-Tesis-Attacks/argus_bots/fa-botnet-capture-20110815-fast-flux-2.csv')
df_b3=pd.read_csv('/ML-Tesis-Attacks/argus_bots/fa-ONOS_Botnet.csv')
df_b4=pd.read_csv('/ML-Tesis-Attacks/argus_bots/fa-S1_Botnet.csv')

In [14]:
df_bts=pd.concat([df_b3, df_b4], axis=0)

In [16]:
df_bts.shape

(102, 123)

In [19]:
df_btst=pd.concat([df_b1,df_b2,df_b3, df_b4], axis=0)

In [20]:
df_btst.shape

(106622, 123)

# Verificacion

In [2]:
#diccionario de acceso a archivos de extraccion de flujos
dir_traficoFA = pl.Path("/ML-Tesis-Attacks/argus_ra/")
"""crear diccionario para el acceso de archivos"""
dicc_traficoFA = {
    'Normal':list(dir_traficoFA.glob('Normal/**/*.csv')),
    'BOTNET': list(dir_traficoFA.glob('botnet/*.csv')),
    'BFA': list(dir_traficoFA.glob('BFA/*.csv')),    
    'Dos': list(dir_traficoFA.glob('DoS/**/*.csv')),  
    'Probe': list(dir_traficoFA.glob('Probe/*.csv')),  
    'WebAttack': list(dir_traficoFA.glob('webattack/*.csv'))  
}

In [3]:
#unir datos por clase
etiquetar_unirdatosdeclase(dicc_traficoFA,'Label', '/ML-Tesis-Attacks/datos_SDN/ArgusDatos/')

En proceso de clase Normal...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (117) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (117) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (117) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)


Proceso completado.... 
 Total de archivos unidos: 22
En proceso de clase BOTNET...
Proceso completado.... 
 Total de archivos unidos: 2
En proceso de clase BFA...
Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase Dos...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)


Proceso completado.... 
 Total de archivos unidos: 18
En proceso de clase Probe...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (26,34,116,117) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata)


Proceso completado.... 
 Total de archivos unidos: 4
En proceso de clase WebAttack...
Proceso completado.... 
 Total de archivos unidos: 6


In [2]:
#importar archivos de datos generados
df_botnetAr=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/ArgusDatos/BOTNET.csv',low_memory=False)
df_dosAr=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/ArgusDatos/Dos.csv',low_memory=False)
df_normalAr=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/ArgusDatos/Normal.csv',low_memory=False)
df_probeAr=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/ArgusDatos/Probe.csv',low_memory=False)
df_bfAr=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/ArgusDatos/BFA.csv',low_memory=False)
df_webAr=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/ArgusDatos/WebAttack.csv',low_memory=False)

In [3]:
#concatenar
datos=pd.concat([df_botnetAr,df_dosAr,df_normalAr,df_probeAr,df_bfAr,df_webAr],axis=0)

In [6]:
df_normalAr.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111720 entries, 0 to 111719
Data columns (total 126 columns):
 #    Column        Dtype  
---   ------        -----  
 0    SrcId         object 
 1    Rank          float64
 2    StartTime     object 
 3    LastTime      object 
 4    SrcStartTime  object 
 5    DstStartTime  object 
 6    SrcLastTime   object 
 7    DstLastTime   object 
 8    Trans         float64
 9    Seq           int64  
 10   Flgs          object 
 11   Dur           float64
 12   SrcDur        float64
 13   DstDur        float64
 14   StdDev        float64
 15   SrcMac        object 
 16   DstMac        object 
 17   SrcAddr       object 
 18   DstAddr       object 
 19   Dir           object 
 20   Proto         object 
 21   Sport         object 
 22   Dport         object 
 23   sTos          float64
 24   dTos          float64
 25   sDSb          object 
 26   dDSb          object 
 27   sCo           float64
 28   dCo           float64
 29   sTtl          

In [5]:
datos.Label.value_counts()

Normal       111720
Probe         51666
Dos           35030
BFA             886
WebAttack       564
BOTNET          193
Name: Label, dtype: int64

In [3]:
"""directorio principal de datos"""
dir_trafico = pl.Path("/ML-Tesis-Attacks/araData/")
"""crear diccionario para el acceso de archivos"""
dicc_trafico = {
    'Normal':list(dir_trafico.glob('Normal/**/*.csv')),    
    'Dos': list(dir_trafico.glob('DoS/**/*.csv')),  
    'Probe': list(dir_trafico.glob('Probe/*.csv'))   
}

In [5]:
"""unir datos por clase"""
etiquetar_unirdatosdeclase(dicc_trafico,'Label', '/ML-Tesis-Attacks/datos_SDN/araData/')

En proceso de clase Normal...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (111) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata, low_memory=False)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (111) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata, low_memory=False)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (111) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata, low_memory=False)


Proceso completado.... 
 Total de archivos unidos: 22
En proceso de clase Dos...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata, low_memory=False)
d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata, low_memory=False)


Proceso completado.... 
 Total de archivos unidos: 18
En proceso de clase Probe...


d:\ML-Tesis-Attacks\Clasiffication_TrafficSDN2\Fase1\libmac.py:37: DtypeWarning: Columns (29,33,110,111) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(archivo_exdata, low_memory=False)


Proceso completado.... 
 Total de archivos unidos: 4


In [6]:
"""importar datos generados"""
df_dos=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/araData/Dos.csv',low_memory=False)
df_normal=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/araData/Normal.csv',low_memory=False)
df_probe=pd.read_csv('/ML-Tesis-Attacks/datos_SDN/araData/Probe.csv',low_memory=False)
"""concatenar datos"""
df_data = pd.concat([df_dos, df_normal, df_probe], axis=0)
"""ver cantidad de muestras por clase"""
df_data.Label.value_counts()

Normal    111719
Probe      49452
Dos        35030
Name: Label, dtype: int64